scripts/cleanup.js

```diff
@@ -53,124 +67,116 @@ function getImages(region, repo, callback) {
+    // Determine whether we still need to delete more images
+    excessImages = excessImages - imagesToDelete.length;
+    if (excessImages <= 0) return callback(null, imagesToDelete);
 
-  if (deployDigests.length > MAX_PRIORITY_IMAGES)
-    digests = digests.concat(deployDigests.slice(0, (deployDigests.length - MAX_PRIORITY_IMAGES)));
+    // Make sure that we always leave at least 50 priority images
+    excessImages = Math.max(
+      Math.min(priority.length - MIN_PRIORITY_IMAGES, excessImages),
+      0
+    );
 
-  console.log('digests ', digests.join(','));
-  return callback(null, digests);
+    // Select the oldest priority images for removal
+    imagesToDelete = imagesToDelete
+      .concat(priority.slice(0, excessImages));
+
+    return callback(null, imagesToDelete);
+  });
 }
 
 module.exports.deleteImages = deleteImages;
-function deleteImages(region, repo, imageIds, callback) {
+function deleteImages(region, repositoryName, images, callback) {
+  const ecr = new AWS.ECR({ region: region });
 
-  let ecr = new AWS.ECR({ region: region });
-  ecr.batchDeleteImage({
-    imageIds: imageIds,
-    repositoryName: repo
-  }, callback);
+  // Must delete images in batches of 100 max
+  const remaining = JSON.parse(JSON.stringify(images));
+  const imageIds = remaining.splice(0, 100);
 
+  if (!imageIds.length) return callback();
+
+  ecr.batchDeleteImage({ imageIds, repositoryName }, (err) => {
+    if (err) return callback(err);
+    deleteImages(region, repositoryName, remaining, callback);
+  });
 }
 
 module.exports.commitType = commitType;
-function commitType(sha, digest, callback) {
-  console.log('sha ', sha);
-  const spawn = require('child_process').spawn;
-
-  let type = {}; type[digest] = '';
-  function shspawn(command, callback) {
-    let cmd = spawn('sh', ['-c', command]);
-    let stdout, stderr;
-    cmd.stderr.on('data', err => {
-      stderr = err.toString('utf-8').trim();
-    });
-    cmd.stdout.on('data', out => {
-      stdout = out.toString('utf-8').trim();
-    });
-    cmd.on('close', () => {
-      // An error is just the git command erroring, which means that the
-      // output of the current command is void for determining type, as
-      // opposed to a true error in the script. So, just return null,
-      // instead of erroring.
-      if (stderr && stderr.length) return callback(null, null);
-      else return callback(null, stdout);
+function commitType(sha, digest, tmpdir, callback) {
+  const result = { digest, type: '' };
+
+  function run(command, callback) {
+    child_process.exec(command, (err, stdout) => {
+      if (err) return callback();
+      callback(null, stdout.trim());
     });
-  } 
+  }
+
+  const merge = `git --git-dir=${tmpdir}/.git cat-file -p ${sha} | grep -Ec '^parent [a-z0-9]{40}'`;
+  const tag = `git --git-dir=${tmpdir}/.git tag | grep ${sha}`;
+  const commit = `git --git-dir=${tmpdir}/.git rev-parse --verify ${sha}`;
```

Comment:
By pulling out these long, complex statements and giving them a clear name, the flow of the rest of the application becomes a lot easier to read.